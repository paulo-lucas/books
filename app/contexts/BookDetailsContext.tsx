import React, {
  createContext,
  PropsWithChildren,
  useCallback,
  useEffect,
  useState,
} from 'react';
import { RouteProp, useRoute, useNavigation } from '@react-navigation/native';
import type { BookDetailsContextData } from '@app/interfaces/bookDetailsContext';
import type { Volume } from '@app/interfaces/volume';
import { useFavorites } from '@app/hooks';
import type { RootStackParamList } from '@app/router/rootStackParams';
import { showVolume } from '@app/services';
import { Alert } from 'react-native';

export const BookDetailsContext = createContext({} as BookDetailsContextData);

export const BookDetailsProvider: React.FC<PropsWithChildren> = ({
  children,
}) => {
  const navigation = useNavigation();
  const {
    params: { id },
  } = useRoute<RouteProp<RootStackParamList, 'Details'>>();

  const [book, setBook] = useState<Volume>();
  const [loading, setLoading] = useState<boolean>(true);
  const { isFavorite, toggleFavorite } = useFavorites();

  const fetchBook = useCallback(async () => {
    if (!id) {
      return navigation.goBack();
    }

    try {
      setLoading(true);
      const { data } = await showVolume(id);
      setBook(data);
      setLoading(false);
    } catch (err) {
      Alert.alert('Could not retrieve book.');
      navigation.goBack();
    }
  }, [id, navigation]);

  useEffect(() => {
    fetchBook();
  }, [fetchBook]);

  const contextData: BookDetailsContextData = {
    book,
    loading,
    isFavorite: () => isFavorite(id),
    toggleFavorite: () => toggleFavorite(id),
  };

  return (
    <BookDetailsContext.Provider value={contextData}>
      {children}
    </BookDetailsContext.Provider>
  );
};
