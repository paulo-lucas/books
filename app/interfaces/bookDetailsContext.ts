import type { Volume } from './volume';

export interface BooksDetailsContextData {
  book: Volume;
  toggleFavorite(): void;
  isFavorite(): boolean;
}
