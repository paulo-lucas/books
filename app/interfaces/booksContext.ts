import type { Volume } from './volume';

export interface BooksState {
  total: number;
  itemsPerPage: number;
  page: number;
  data: Array<Volume>;
}

export interface BooksContextData {
  searchText: string;
  onChangeSearch(text: string): void;
  books: BooksState;
  refreshing: boolean;
  toggleOrderBy(): void;
  toggleFavoriteFilter(): void;
  toggleFavorite(isbn: string): void;
  isFavorite(favorite: string): boolean;
  onClearSearch(): void;
  refreshBooks(page?: number): void;
}

export interface UpdatePayload {
  type: 'update';
  page?: number;
  total: number;
  data: Array<Volume>;
}

export interface ClearPayload {
  type: 'clear';
}
