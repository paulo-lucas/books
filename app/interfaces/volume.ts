export interface VolumeSearchRequest {
  q: VolumeSearchQuery;
  startIndex: number;
  maxResults: number;
  orderBy: 'newest' | 'relevance';
}

export interface VolumeSearchResponse {
  kind: string;
  totalItems: number;
  items: Array<Volume>;
}

export interface VolumeSearchQuery {
  search?: string;
  title?: string;
  author?: string;
  publisher?: string;
  isbn?: Array<string>;
}

export interface VolumeInfo {
  title: string;
  authors: Array<String>;
  publisher: string;
  publishedDate: string;
  description: string;
  pageCount: number;
  categories: Array<string>;
  maturityRating: string;
  imageLinks: { smallThumbnail: string; thumbnail: string };
  language: string;
  previewLink: string;
  infoLink: string;
  canonicalVolumeLink: string;
}

export interface SaleInfo {
  country: string;
  saleability: string;
  isEbook: boolean;
}

export interface SearchInfo {
  textSnippet: string;
}

export interface Volume {
  kind: string;
  id: string;
  selfLink: string;
  volumeInfo: VolumeInfo;
  saleInfo: SaleInfo;
  accessInfo: any;
  searchInfo: SearchInfo;
}
