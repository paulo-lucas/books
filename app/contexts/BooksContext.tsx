import React, {
  createContext,
  useState,
  useReducer,
  PropsWithChildren,
  useEffect,
} from 'react';
import { BooksContextData } from '@app/interfaces/booksContext';
import { booksReducer, booksInitialState } from '@app/states/BooksReducer';
import { useDebounce } from '@app/hooks';
import { searchVolumes, storeFavorites, getFavorites } from '@app/services';
import { Alert } from 'react-native';
import { useCallback } from 'react';

export const BooksContext = createContext({} as BooksContextData);

export const BooksProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [books, dispatch] = useReducer(booksReducer, booksInitialState);
  const [searchText, setSearchText] = useState<string>('');
  const [refreshing, setRefreshing] = useState<boolean>(false);
  const [orderBy, setOrderBy] = useState<'relevance' | 'newest'>('relevance');
  const [favorites, setFavorites] = useState<Array<string>>([]);
  const [filterByFavorites, setFilterByFavorites] = useState<boolean>(false);

  const fetchFavorites = useCallback(async () => {
    const storageFavorites = await getFavorites();
    setFavorites(storageFavorites);
  }, []);

  const onClearSearch = () => dispatch({ type: 'clear' });

  const toggleFavoriteFilter = () => setFilterByFavorites(!filterByFavorites);

  const toggleOrderBy = () =>
    setOrderBy(orderBy === 'newest' ? 'relevance' : 'newest');

  const isFavorite = (isbn: string) => favorites.includes(isbn);

  const toggleFavorite = (isbn: string) => {
    let updatedFavorites;

    if (favorites.includes(isbn)) {
      updatedFavorites = favorites.filter(fav => fav !== isbn);
    } else {
      updatedFavorites = [...favorites, isbn];
    }

    setFavorites(updatedFavorites);
    storeFavorites(updatedFavorites);
  };

  const fetchBooks = useCallback(
    async (page: number, searchFromDebounce?: string) => {
      if (!searchText && !searchFromDebounce) {
        return onClearSearch();
      }

      const search = searchFromDebounce ?? searchText;
      const isbn = filterByFavorites ? favorites : [];
      const q = { search, isbn };
      const { itemsPerPage: maxResults } = books;
      const startIndex = page * maxResults;

      try {
        setRefreshing(true);
        const { data } = await searchVolumes({
          q,
          startIndex,
          maxResults,
          orderBy,
        });

        dispatch({
          type: 'update',
          total: data.totalItems,
          data: data.items ?? [],
        });

        setRefreshing(false);
      } catch (err) {
        Alert.alert('Could not retrieve books.');
      }
    },
    [books, favorites, filterByFavorites, orderBy, searchText],
  );

  const debounceSearch = useDebounce(fetchBooks, 1000);

  const onChangeSearch = (text: string) => {
    setSearchText(text);
    debounceSearch(0, text);
  };

  const refreshBooks = (page: number = 0) => {
    fetchBooks(page);
  };

  useEffect(() => {
    fetchFavorites();
  }, [fetchFavorites]);

  const contextData: BooksContextData = {
    books,
    refreshBooks,
    refreshing,
    searchText,
    onChangeSearch,
    onClearSearch,
    toggleOrderBy,
    toggleFavoriteFilter,
    isFavorite,
    toggleFavorite,
  };

  return (
    <BooksContext.Provider value={contextData}>
      {children}
    </BooksContext.Provider>
  );
};
