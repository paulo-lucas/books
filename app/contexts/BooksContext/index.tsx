import React, {
  createContext,
  useState,
  useReducer,
  PropsWithChildren,
  useEffect,
} from 'react';
import { BooksContextData } from '@app/interfaces/booksContext';
import { booksReducer, booksInitialState } from '@app/states/BooksReducer';
import { useDebounce } from '@app/hooks';
import { searchVolumes, storeFavorites, getFavorites } from '@app/services';
import { Alert } from 'react-native';
import { useCallback } from 'react';
import { VolumeSearchRequest } from '@app/interfaces/volume';

export const BooksContext = createContext({} as BooksContextData);

export const BooksProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [books, dispatch] = useReducer(booksReducer, booksInitialState);
  const [searchText, setSearchText] = useState<string>('');
  const [searchTextDebounce, setSearchTextDebounce] = useState<string>('');
  const [refreshing, setRefreshing] = useState<boolean>(false);
  const [orderBy, setOrderBy] = useState<'relevance' | 'newest'>('relevance');
  const [favorites, setFavorites] = useState<Array<string>>([]);
  const [filterByFavorites, setFilterByFavorites] = useState<boolean>(false);

  const fetchFavorites = useCallback(async () => {
    const favoritesFromStorage = await getFavorites();
    setFavorites(favoritesFromStorage);
  }, []);

  // Handle every request to list books, based on states
  const fetchBooks = useCallback(async () => {
    if (!searchTextDebounce && (!filterByFavorites || !favorites.length)) {
      return onClearSearch();
    }

    const identifier = filterByFavorites ? favorites : [];
    const q = { search: searchTextDebounce, identifier };
    const { itemsPerPage: maxResults, page } = books;
    const startIndex = page * maxResults;

    try {
      const request: VolumeSearchRequest = {
        q,
        startIndex,
        maxResults,
        orderBy,
        filter: 'partial',
        printType: 'books',
      };

      if (JSON.stringify(request) === books.request) {
        return;
      }

      setRefreshing(true);
      const { data } = await searchVolumes(request);

      const items = filterByFavorites
        ? data.items?.filter(item => favorites.includes(item.id))
        : data.items;

      dispatch({
        type: 'update',
        total: data.items?.length ?? 0,
        distinctTotal: items.length,
        data: items ?? [],
        page: page,
        request: JSON.stringify(request),
      });

      setRefreshing(false);
    } catch (err) {
      onClearSearch();
      setFilterByFavorites(false);
      Alert.alert('Could not retrieve books.');
    }
  }, [books, favorites, filterByFavorites, orderBy, searchTextDebounce]);

  const onClearSearch = () => dispatch({ type: 'clear' });

  const toggleOrderBy = () =>
    setOrderBy(orderBy === 'newest' ? 'relevance' : 'newest');

  // Toggle the filter, will make table show only favorited books
  const toggleFavoriteFilter = () => setFilterByFavorites(!filterByFavorites);

  // Toggle single book favorited / not favorited
  const toggleFavorite = (identifier: string) => {
    let updatedFavorites;

    if (favorites.includes(identifier)) {
      updatedFavorites = favorites.filter(fav => fav !== identifier);
    } else {
      updatedFavorites = [...favorites, identifier];
    }

    updatedFavorites = updatedFavorites.filter(fav => !!fav);

    setFavorites(updatedFavorites);
    storeFavorites(updatedFavorites);
  };

  const isFavorite = (identifier: string) => favorites.includes(identifier);

  const debounceSearch = useDebounce(setSearchTextDebounce, 500);

  const onChangeSearch = (text: string) => {
    setSearchText(text);
    debounceSearch(text);
  };

  const onChangePage = (page: number) => dispatch({ type: 'moveToPage', page });

  useEffect(() => {
    fetchBooks();
  }, [fetchBooks]);

  useEffect(() => {
    fetchFavorites();
  }, [fetchFavorites]);

  const contextData: BooksContextData = {
    books,
    refreshing,
    searchText,
    onChangeSearch,
    onClearSearch,
    fetchBooks,
    toggleOrderBy,
    toggleFavoriteFilter,
    orderBy,
    filterByFavorites,
    isFavorite,
    toggleFavorite,
    onChangePage,
  };

  return (
    <BooksContext.Provider value={contextData}>
      {children}
    </BooksContext.Provider>
  );
};
