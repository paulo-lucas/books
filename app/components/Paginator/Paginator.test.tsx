import 'react-native';
import React from 'react';
import { render, screen } from '../../../tests/test-utils';
import { Paginator } from './';
import { BooksContext } from '@app/contexts/BooksContext';
import { BooksContextData } from '@app/interfaces/booksContext';

const mockPage = 2;
const mockTotal = 15;
const mockItemsPerPage = 20;

const onChangePage = jest.fn();

const renderPaginator = ({ refreshing = false }) => {
  const books = {
    page: mockPage,
    total: mockTotal,
    itemsPerPage: mockItemsPerPage,
  };

  return render(
    <BooksContext.Provider
      value={
        { books, onChangePage, refreshing } as unknown as BooksContextData
      }>
      <Paginator />
    </BooksContext.Provider>,
  );
};

describe('paginator', () => {
  it('should render correctly', () => {
    renderPaginator({});
    expect(screen.toJSON()).toMatchSnapshot('paginator');
  });

  it('should render correctly when refreshing', () => {
    renderPaginator({ refreshing: true });
    expect(screen.toJSON()).toMatchSnapshot('paginator refreshing');
  });

  it('should display page and item range', () => {
    renderPaginator({});

    const offset = mockPage * mockItemsPerPage;

    expect(screen.getByText(`Page ${mockPage + 1}`)).toBeTruthy();
    expect(
      screen.getByText(`Items ${offset + 1}-${offset + mockTotal}`),
    ).toBeTruthy();
  });
});
