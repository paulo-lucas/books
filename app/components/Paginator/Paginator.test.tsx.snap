// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`paginator should render correctly when refreshing: paginator refreshing 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    Object {
      "backgroundColor": undefined,
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-evenly",
        "paddingBottom": 20,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 2,
          },
          undefined,
          false,
        ]
      }
      testID="RNE_BUTTON_WRAPPER"
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "busy": false,
            "disabled": true,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "#00DAC4",
              "borderRadius": 2,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
              "paddingHorizontal": 10,
            }
          }
        >
          <View
            style={
              Object {
                "overflow": "hidden",
              }
            }
            testID="RNE__ICON__CONTAINER"
          >
            <View
              accessibilityRole="button"
              android_ripple={
                Object {
                  "borderless": false,
                  "color": "rgba(0, 0, 0, 0.3)",
                  "radius": -5,
                }
              }
              disabled={false}
              replaceTheme={[Function]}
              testID="RNE__ICON__CONTAINER_ACTION"
              updateTheme={[Function]}
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
                testID="RNE__ICON"
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#86939e",
                        "fontSize": 24,
                      },
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                  testID="RNE__ICON__Component"
                >
                  
                </Text>
              </View>
            </View>
          </View>
          <Text
            style={
              Object {
                "color": "hsl(208, 8%, 63%)",
                "fontSize": 18,
                "paddingVertical": 1,
                "textAlign": "center",
              }
            }
          >
            Prev
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#242424",
          }
        }
      >
        Page 
        3
      </Text>
      <Text
        style={
          Object {
            "color": "#242424",
          }
        }
      >
        Items 
        41
        -
        55
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 2,
          },
          undefined,
          false,
        ]
      }
      testID="RNE_BUTTON_WRAPPER"
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "busy": false,
            "disabled": true,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "#00DAC4",
              "borderRadius": 2,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
              "paddingHorizontal": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "hsl(208, 8%, 63%)",
                "fontSize": 18,
                "paddingVertical": 1,
                "textAlign": "center",
              }
            }
          >
            Next
          </Text>
          <View
            style={
              Object {
                "overflow": "hidden",
              }
            }
            testID="RNE__ICON__CONTAINER"
          >
            <View
              accessibilityRole="button"
              android_ripple={
                Object {
                  "borderless": false,
                  "color": "rgba(0, 0, 0, 0.3)",
                  "radius": -5,
                }
              }
              disabled={false}
              replaceTheme={[Function]}
              testID="RNE__ICON__CONTAINER_ACTION"
              updateTheme={[Function]}
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
                testID="RNE__ICON"
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#86939e",
                        "fontSize": 24,
                      },
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                  testID="RNE__ICON__Component"
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;

exports[`paginator should render correctly: paginator 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    Object {
      "backgroundColor": undefined,
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-evenly",
        "paddingBottom": 20,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 2,
          },
          undefined,
          false,
        ]
      }
      testID="RNE_BUTTON_WRAPPER"
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "busy": false,
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "#00DAC4",
              "borderRadius": 2,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
              "paddingHorizontal": 10,
            }
          }
        >
          <View
            style={
              Object {
                "overflow": "hidden",
              }
            }
            testID="RNE__ICON__CONTAINER"
          >
            <View
              accessibilityRole="button"
              android_ripple={
                Object {
                  "borderless": false,
                  "color": "rgba(0, 0, 0, 0.3)",
                  "radius": -5,
                }
              }
              disabled={false}
              replaceTheme={[Function]}
              testID="RNE__ICON__CONTAINER_ACTION"
              updateTheme={[Function]}
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
                testID="RNE__ICON"
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#2089dc",
                        "fontSize": 24,
                      },
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                  testID="RNE__ICON__Component"
                >
                  
                </Text>
              </View>
            </View>
          </View>
          <Text
            style={
              Object {
                "color": "#00DAC4",
                "fontSize": 18,
                "paddingVertical": 1,
                "textAlign": "center",
              }
            }
          >
            Prev
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#242424",
          }
        }
      >
        Page 
        3
      </Text>
      <Text
        style={
          Object {
            "color": "#242424",
          }
        }
      >
        Items 
        41
        -
        55
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 2,
          },
          undefined,
          false,
        ]
      }
      testID="RNE_BUTTON_WRAPPER"
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "busy": false,
            "disabled": true,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "#00DAC4",
              "borderRadius": 2,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
              "paddingHorizontal": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "hsl(208, 8%, 63%)",
                "fontSize": 18,
                "paddingVertical": 1,
                "textAlign": "center",
              }
            }
          >
            Next
          </Text>
          <View
            style={
              Object {
                "overflow": "hidden",
              }
            }
            testID="RNE__ICON__CONTAINER"
          >
            <View
              accessibilityRole="button"
              android_ripple={
                Object {
                  "borderless": false,
                  "color": "rgba(0, 0, 0, 0.3)",
                  "radius": -5,
                }
              }
              disabled={false}
              replaceTheme={[Function]}
              testID="RNE__ICON__CONTAINER_ACTION"
              updateTheme={[Function]}
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
                testID="RNE__ICON"
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#86939e",
                        "fontSize": 24,
                      },
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                  testID="RNE__ICON__Component"
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
